Cython/Compiler/Code.py

```diff
@@ -2060,131 +2062,136 @@ def entry_as_pyobject(self, entry):
         type = entry.type
         if (not entry.is_self_arg and not entry.type.is_complete()
             or entry.type.is_extension_type):
-            return "(PyObject *)" + entry.cname
+            return "(PyObject *)" + entry.cname # FIXME need to implement this
         else:
             return entry.cname
 
     def as_pyobject(self, cname, type):
         from .PyrexTypes import py_object_type, typecast
         return typecast(py_object_type, type, cname)
 
-    def put_gotref(self, cname):
-        self.putln("__Pyx_GOTREF(%s);" % cname)
+    def put_gotref(self, cname, type):
+        self.putln(type.generate_gotref(cname))
+
+    def put_exprnode_gotref(self, node):
+        if node.is_temp and node.type.is_pyobject:
+            # temp py_objects are always just py_object_type
+            # not an special type
+            from .PyrexTypes import py_object_type as type
```

Comment:
Careful. I'd rather not let `Code.py` know about types or expressions. (I know there's a bit of code in there using types already, but that's not ideal.)

Rather, I think an `ExprNode` can know how to generate a giveref/gotref/etc. for itself given a `code` generator object, i.e. `def generate_gotref(self, code): …` etc.

Reply:
To clarify: something like `type.do_something()` is fine, a typecast is limit, processing nodes is way over the line. Nodes use a code writer, not the other way round.

Reply:
Yes, that does make more sense.

I'm keeping a separate branch with the reference counting changes which I'll put in it's own pull request when this+it is complete.